File: protocol/rust/src/json_convert.rs
//! JSON conversion for LSE — bidirectional SemanticNode <-> JSON.

use std::collections::HashMap;

use serde_json::Value;

use crate::types::*;

/// A diagnostic from JSON validation.
#[derive(Debug, Clone)]
pub struct Diagnostic {
    pub severity: String,
    pub code: String,
    pub message: String,
}

/// Valid value types for JSON validation.
const VALID_VALUE_TYPES: &[&str] = &[
    "selector",
    "literal",
    "reference",
    "expression",
    "property-path",
    "flag",
];

/// Validate a SemanticJSON object. Returns a list of diagnostics (empty = valid).
pub fn validate_json(data: &Value) -> Vec<Diagnostic> {
    let mut diagnostics = Vec::new();

    let obj = match data.as_object() {
        Some(o) => o,
        None => {
            diagnostics.push(Diagnostic {
                severity: "error".to_string(),
                code: "INVALID_ACTION".to_string(),
                message: "Missing or invalid 'action' field (must be a non-empty string)"
                    .to_string(),
            });
            return diagnostics;
        }
    };

    // Check action
    match obj.get("action").and_then(|a| a.as_str()) {
        Some(s) if !s.is_empty() => {}
        _ => {
            diagnostics.push(Diagnostic {
                severity: "error".to_string(),
                code: "INVALID_ACTION".to_string(),
                message: "Missing or invalid 'action' field (must be a non-empty string)"
                    .to_string(),
            });
        }
    }

    // Check roles
    if let Some(roles) = obj.get("roles").and_then(|r| r.as_object()) {
        for (role_name, role_value) in roles {
            let rv = match role_value.as_object() {
                Some(o) => o,
                None => {
                    diagnostics.push(Diagnostic {
                        severity: "error".to_string(),
                        code: "INVALID_ROLE_VALUE".to_string(),
                        message: format!(
                            "Role \"{}\" must be an object with type and value",
                            role_name
                        ),
                    });
                    continue;
                }
            };

            let vtype = rv.get("type").and_then(|t| t.as_str()).unwrap_or("");
            if !VALID_VALUE_TYPES.contains(&vtype) {
                diagnostics.push(Diagnostic {
                    severity: "error".to_string(),
                    code: "INVALID_VALUE_TYPE".to_string(),
                    message: format!("Role \"{}\" has invalid type \"{}\"", role_name, vtype),
                });
            }

            if !rv.contains_key("value") && vtype != "expression" {
                if !rv.contains_key("raw") {
                    diagnostics.push(Diagnostic {
                        severity: "error".to_string(),
                        code: "MISSING_VALUE".to_string(),
                        message: format!("Role \"{}\" is missing value field", role_name),
                    });
                }
            }
        }
    }

    // Check trigger
    if let Some(trigger) = obj.get("trigger") {
        let event = trigger
            .as_object()
            .and_then(|t| t.get("event"))
            .and_then(|e| e.as_str());
        match event {
            Some(s) if !s.is_empty() => {}
            _ => {
                diagnostics.push(Diagnostic {
                    severity: "error".to_string(),
                    code: "INVALID_TRIGGER".to_string(),
                    message: "Trigger must have a non-empty 'event' string".to_string(),
                });
            }
        }
    }

    diagnostics
}

/// Convert a SemanticJSON Value to a SemanticNode.
///
/// Accepts both full-fidelity and LLM-simplified formats.
pub fn from_json(data: &Value) -> Result<SemanticNode, String> {
    let obj = data
        .as_object()
        .ok_or_else(|| "Expected JSON object".to_string())?;

    let action = obj
        .get("action")
        .and_then(|a| a.as_str())
        .unwrap_or("")
        .to_string();

    let raw_roles = obj
        .get("roles")
        .and_then(|r| r.as_object())
        .cloned()
        .unwrap_or_default();

    let mut roles: HashMap<String, SemanticValue> = HashMap::new();
    for (role_name, role_value) in &raw_roles {
        if role_value.is_object() {
            roles.insert(role_name.clone(), convert_json_value(role_value));
        }
    }

    // If trigger present, wrap in event handler
    if let Some(trigger) = obj.get("trigger").and_then(|t| t.as_object()) {
        let event_name = trigger
            .get("event")
            .and_then(|e| e.as_str())
            .unwrap_or("")
            .to_string();
        let mut event_roles: HashMap<String, SemanticValue> = HashMap::new();
        event_roles.insert(
            "event".to_string(),
            literal_value(DynValue::String(event_name), "string"),
        );
        let body_node = SemanticNode::command(&action, roles);
        return Ok(SemanticNode::event_handler(event_roles, vec![body_node]));
    }

    // Check for full-fidelity format with kind
    let kind = obj
        .get("kind")
        .and_then(|k| k.as_str())
        .unwrap_or("command");

    if kind == "event-handler" {
        let body_data = obj
            .get("body")
            .and_then(|b| b.as_array())
            .cloned()
            .unwrap_or_default();
        let mut body = Vec::new();
        for b in &body_data {
            body.push(from_json(b)?);
        }
        return Ok(SemanticNode {
            kind: NodeKind::EventHandler,
            action,
            roles,
            body,
            statements: Vec::new(),
            chain_type: None,
        });
    }

    if kind == "compound" {
        let stmts_data = obj
            .get("statements")
            .and_then(|s| s.as_array())
            .cloned()
            .unwrap_or_default();
        let mut stmts = Vec::new();
        for s in &stmts_data {
            stmts.push(from_json(s)?);
        }
        let chain_type = obj
            .get("chainType")
            .and_then(|c| c.as_str())
            .unwrap_or("then")
            .to_string();
        return Ok(SemanticNode {
            kind: NodeKind::Compound,
            action,
            roles,
            body: Vec::new(),
            statements: stmts,
            chain_type: Some(chain_type),
        });
    }

    Ok(SemanticNode::command(&action, roles))
}

/// Convert a SemanticNode to a full-fidelity JSON Value.
pub fn to_json(node: &SemanticNode) -> Value {
    serde_json::to_value(node).unwrap_or(Value::Null)
}

/// Convert a JSON value object to a SemanticValue.
fn convert_json_value(data: &Value) -> SemanticValue {
    let obj = match data.as_object() {
        Some(o) => o,
        None => return literal_value(DynValue::String(String::new()), "string"),
    };

    let vtype = obj
        .get("type")
        .and_then(|t| t.as_str())
        .unwrap_or("literal");
    let raw_value = obj.get("value").or_else(|| obj.get("raw"));

    match vtype {
        "selector" => {
            let v = raw_value
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string();
            selector_value(&v)
        }
        "literal" => {
            if let Some(val) = raw_value {
                if val.is_boolean() {
                    literal_value(DynValue::Bool(val.as_bool().unwrap()), "boolean")
                } else if val.is_number() {
                    let dv = DynValue::from_json_value(val);
                    literal_value(dv, "number")
                } else {
                    let data_type = obj
                        .get("dataType")
                        .and_then(|dt| dt.as_str())
                        .unwrap_or("string");
                    let s = val.as_str().unwrap_or("").to_string();
                    literal_value(DynValue::String(s), data_type)
                }
            } else {
                literal_value(DynValue::String(String::new()), "string")
            }
        }
        "reference" => {
            let v = raw_value
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string();
            reference_value(&v)
        }
        "expression" => {
            let r = obj
                .get("raw")
                .or(raw_value)
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string();
            expression_value(&r)
        }
        "flag" => {
            let name = obj
                .get("name")
                .and_then(|n| n.as_str())
                .or_else(|| raw_value.and_then(|v| v.as_str()))
                .unwrap_or("")
                .to_string();
            let enabled = obj
                .get("enabled")
                .and_then(|e| e.as_bool())
                .or_else(|| raw_value.and_then(|v| v.as_bool()))
                .unwrap_or(true);
            flag_value(&name, enabled)
        }
        "property-path" => {
            let r = raw_value
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string();
            expression_value(&r)
        }
        _ => {
            let s = raw_value
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string();
            literal_value(DynValue::String(s), "string")
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn test_validate_json_valid() {
        let data = json!({
            "action": "toggle",
            "roles": {
                "patient": {"type": "selector", "value": ".active"}
            }
        });
        let diags = validate_json(&data);
        assert!(diags.is_empty(), "got: {:?}", diags);
    }

    #[test]
    fn test_validate_json_missing_action() {
        let data = json!({"roles": {}});
        let diags = validate_json(&data);
        assert_eq!(diags.len(), 1);
        assert_eq!(diags[0].code, "INVALID_ACTION");
    }

    #[test]
    fn test_validate_json_invalid_role_value() {
        let data = json!({
            "action": "toggle",
            "roles": {"patient": "not-an-object"}
        });
        let diags = validate_json(&data);
        assert!(diags.iter().any(|d| d.code == "INVALID_ROLE_VALUE"));
    }

    #[test]
    fn test_validate_json_invalid_value_type() {
        let data = json!({
            "action": "toggle",
            "roles": {"patient": {"type": "unknown", "value": ".active"}}
        });
        let diags = validate_json(&data);
        assert!(diags.iter().any(|d| d.code == "INVALID_VALUE_TYPE"));
    }

    #[test]
    fn test_validate_json_missing_value() {
        let data = json!({
            "action": "toggle",
            "roles": {"patient": {"type": "selector"}}
        });
        let diags = validate_json(&data);
        assert!(diags.iter().any(|d| d.code == "MISSING_VALUE"));
    }

    #[test]
    fn test_validate_json_valid_trigger() {
        let data = json!({
            "action": "toggle",
            "roles": {"patient": {"type": "selector", "value": ".active"}},
            "trigger": {"event": "click"}
        });
        let diags = validate_json(&data);
        assert!(diags.is_empty(), "got: {:?}", diags);
    }

    #[test]
    fn test_validate_json_invalid_trigger() {
        let data = json!({
            "action": "toggle",
            "roles": {"patient": {"type": "selector", "value": ".active"}},
            "trigger": {"event": ""}
        });
        let diags = validate_json(&data);
        assert!(diags.iter().any(|d| d.code == "INVALID_TRIGGER"));
    }

    #[test]
    fn test_validate_json_flag_type() {
        let data = json!({
            "action": "column",
            "roles": {"primary-key": {"type": "flag", "value": true}}
        });
        let diags = validate_json(&data);
        assert!(diags.is_empty(), "got: {:?}", diags);
    }

    #[test]
    fn test_from_json_basic_command() {
        let data = json!({
            "action": "toggle",
            "roles": {"patient": {"type": "selector", "value": ".active"}}
        });
        let node = from_json(&data).unwrap();
        assert_eq!(node.kind, NodeKind::Command);
        assert_eq!(node.action, "toggle");
        assert_eq!(node.roles["patient"].value_type, ValueType::Selector);
    }

    #[test]
    fn test_from_json_trigger_wraps_event_handler() {
        let data = json!({
            "action": "toggle",
            "roles": {"patient": {"type": "selector", "value": ".active"}},
            "trigger": {"event": "click"}
        });
        let node = from_json(&data).unwrap();
        assert_eq!(node.kind, NodeKind::EventHandler);
        assert_eq!(node.action, "on");
        assert_eq!(node.body.len(), 1);
        assert_eq!(node.body[0].action, "toggle");
    }

    #[test]
    fn test_from_json_full_fidelity_event_handler() {
        let data = json!({
            "kind": "event-handler",
            "action": "on",
            "roles": {
                "event": {"type": "literal", "value": "click", "dataType": "string"}
            },
            "body": [{
                "kind": "command",
                "action": "toggle",
                "roles": {
                    "patient": {"type": "selector", "value": ".active"}
                }
            }]
        });
        let node = from_json(&data).unwrap();
        assert_eq!(node.kind, NodeKind::EventHandler);
        assert_eq!(node.body.len(), 1);
    }

    #[test]
    fn test_from_json_compound_node() {
        let data = json!({
            "kind": "compound",
            "action": "compound",
            "chainType": "then",
            "statements": [
                {
                    "kind": "command",
                    "action": "add",
                    "roles": {"patient": {"type": "selector", "value": ".loading"}}
                },
                {
                    "kind": "command",
                    "action": "fetch",
                    "roles": {"source": {"type": "literal", "value": "/api/data", "dataType": "string"}}
                }
            ]
        });
        let node = from_json(&data).unwrap();
        assert_eq!(node.kind, NodeKind::Compound);
        assert_eq!(node.chain_type.as_deref(), Some("then"));
        assert_eq!(node.statements.len(), 2);
    }

    #[test]
    fn test_to_json_basic_command() {
        let mut roles = HashMap::new();
        roles.insert("patient".to_string(), selector_value(".active"));
        let node = SemanticNode::command("toggle", roles);
        let result = to_json(&node);
        assert_eq!(result["action"], "toggle");
        assert_eq!(result["roles"]["patient"]["type"], "selector");
        assert_eq!(result["roles"]["patient"]["value"], ".active");
    }

    #[test]
    fn test_to_json_flag() {
        let mut roles = HashMap::new();
        roles.insert(
            "name".to_string(),
            literal_value(DynValue::String("id".to_string()), "string"),
        );
        roles.insert("primary-key".to_string(), flag_value("primary-key", true));
        let node = SemanticNode::command("column", roles);
        let result = to_json(&node);
        let pk = &result["roles"]["primary-key"];
        assert_eq!(pk["type"], "flag");
        assert_eq!(pk["name"], "primary-key");
        assert_eq!(pk["enabled"], true);
    }

    #[test]
    fn test_round_trip_json() {
        let original = json!({
            "action": "toggle",
            "roles": {
                "patient": {"type": "selector", "value": ".active"},
                "destination": {"type": "selector", "value": "#button"}
            }
        });
        let node = from_json(&original).unwrap();
        let result = to_json(&node);
        assert_eq!(result["action"], "toggle");
        assert_eq!(result["roles"]["patient"]["type"], "selector");
        assert_eq!(result["roles"]["patient"]["value"], ".active");
    }
}

File: protocol/go/json_convert.go
package lse

import (
	"fmt"
)

// Diagnostic represents a validation error or warning.
type Diagnostic struct {
	Severity string `json:"severity"`
	Code     string `json:"code"`
	Message  string `json:"message"`
}

// ValidateJSON validates a SemanticJSON map. Returns a slice of diagnostics (empty = valid).
func ValidateJSON(data map[string]any) []Diagnostic {
	var diags []Diagnostic

	// Check action
	action, _ := data["action"].(string)
	if action == "" {
		diags = append(diags, Diagnostic{
			Severity: "error",
			Code:     "INVALID_ACTION",
			Message:  "Missing or invalid 'action' field (must be a non-empty string)",
		})
	}

	// Check roles
	if roles, ok := data["roles"].(map[string]any); ok {
		for roleName, rv := range roles {
			roleValue, ok := rv.(map[string]any)
			if !ok {
				diags = append(diags, Diagnostic{
					Severity: "error",
					Code:     "INVALID_ROLE_VALUE",
					Message:  fmt.Sprintf("Role %q must be an object with type and value", roleName),
				})
				continue
			}

			vtype, _ := roleValue["type"].(string)
			validTypes := map[string]bool{
				"selector": true, "literal": true, "reference": true,
				"expression": true, "property-path": true, "flag": true,
			}
			if !validTypes[vtype] {
				diags = append(diags, Diagnostic{
					Severity: "error",
					Code:     "INVALID_VALUE_TYPE",
					Message:  fmt.Sprintf("Role %q has invalid type %q", roleName, vtype),
				})
			}

			_, hasValue := roleValue["value"]
			_, hasRaw := roleValue["raw"]
			if !hasValue && vtype != "expression" && !hasRaw {
				diags = append(diags, Diagnostic{
					Severity: "error",
					Code:     "MISSING_VALUE",
					Message:  fmt.Sprintf("Role %q is missing value field", roleName),
				})
			}
		}
	}

	// Check trigger
	if trigger, ok := data["trigger"].(map[string]any); ok {
		event, _ := trigger["event"].(string)
		if event == "" {
			diags = append(diags, Diagnostic{
				Severity: "error",
				Code:     "INVALID_TRIGGER",
				Message:  "Trigger must have a non-empty 'event' string",
			})
		}
	}

	return diags
}

// FromJSON converts a SemanticJSON map to a SemanticNode.
// Accepts both full-fidelity and LLM-simplified formats.
func FromJSON(data map[string]any) (*SemanticNode, error) {
	action, _ := data["action"].(string)
	rawRoles, _ := data["roles"].(map[string]any)

	roles := make(map[string]SemanticValue)
	for roleName, rv := range rawRoles {
		if roleMap, ok := rv.(map[string]any); ok {
			roles[roleName] = convertJSONValue(roleMap)
		}
	}

	// If trigger present, wrap in event handler
	if trigger, ok := data["trigger"].(map[string]any); ok {
		eventName, _ := trigger["event"].(string)
		eventRoles := map[string]SemanticValue{
			"event": LiteralValue(eventName, "string"),
		}
		bodyNode := SemanticNode{
			Kind:   KindCommand,
			Action: action,
			Roles:  roles,
		}
		return &SemanticNode{
			Kind:   KindEventHandler,
			Action: "on",
			Roles:  eventRoles,
			Body:   []SemanticNode{bodyNode},
		}, nil
	}

	// Check for full-fidelity format with kind
	kind := NodeKind("command")
	if k, ok := data["kind"].(string); ok && k != "" {
		kind = NodeKind(k)
	}

	if kind == KindEventHandler {
		var body []SemanticNode
		if bodyData, ok := data["body"].([]any); ok {
			for _, b := range bodyData {
				if bMap, ok := b.(map[string]any); ok {
					bodyNode, err := FromJSON(bMap)
					if err != nil {
						return nil, err
					}
					body = append(body, *bodyNode)
				}
			}
		}
		ehNode := &SemanticNode{
			Kind:   KindEventHandler,
			Action: action,
			Roles:  roles,
			Body:   body,
		}
		// v1.2: annotations on event-handler
		if anns := deserializeAnnotations(data["annotations"]); anns != nil {
			ehNode.Annotations = anns
		}
		return ehNode, nil
	}

	if kind == KindCompound {
		var stmts []SemanticNode
		if stmtsData, ok := data["statements"].([]any); ok {
			for _, s := range stmtsData {
				if sMap, ok := s.(map[string]any); ok {
					stmtNode, err := FromJSON(sMap)
					if err != nil {
						return nil, err
					}
					stmts = append(stmts, *stmtNode)
				}
			}
		}
		chainType, _ := data["chainType"].(string)
		if chainType == "" {
			chainType = "then"
		}
		cmpNode := &SemanticNode{
			Kind:       KindCompound,
			Action:     action,
			Roles:      roles,
			Statements: stmts,
			ChainType:  chainType,
		}
		// v1.2: annotations on compound
		if anns := deserializeAnnotations(data["annotations"]); anns != nil {
			cmpNode.Annotations = anns
		}
		return cmpNode, nil
	}

	node := &SemanticNode{
		Kind:   KindCommand,
		Action: action,
		Roles:  roles,
	}

	// Deserialize v1.1 conditional fields
	if thenData, ok := data["thenBranch"].([]any); ok {
		for _, t := range thenData {
			if tMap, ok := t.(map[string]any); ok {
				tNode, err := FromJSON(tMap)
				if err != nil {
					return nil, err
				}
				node.ThenBranch = append(node.ThenBranch, *tNode)
			}
		}
	}
	if elseData, ok := data["elseBranch"].([]any); ok {
		for _, e := range elseData {
			if eMap, ok := e.(map[string]any); ok {
				eNode, err := FromJSON(eMap)
				if err != nil {
					return nil, err
				}
				node.ElseBranch = append(node.ElseBranch, *eNode)
			}
		}
	}

	// Deserialize v1.1 loop fields
	if lv, ok := data["loopVariant"].(string); ok {
		node.LoopVariant = lv
	}
	if loopData, ok := data["loopBody"].([]any); ok {
		for _, l := range loopData {
			if lMap, ok := l.(map[string]any); ok {
				lNode, err := FromJSON(lMap)
				if err != nil {
					return nil, err
				}
				node.LoopBody = append(node.LoopBody, *lNode)
			}
		}
	}
	if lvar, ok := data["loopVariable"].(string); ok {
		node.LoopVariable = lvar
	}
	if ivar, ok := data["indexVariable"].(string); ok {
		node.IndexVariable = ivar
	}

	// Deserialize v1.2 command body (used by try, all, race)
	if bodyData, ok := data["body"].([]any); ok {
		for _, b := range bodyData {
			if bMap, ok := b.(map[string]any); ok {
				bNode, err := FromJSON(bMap)
				if err != nil {
					return nil, err
				}
				node.Body = append(node.Body, *bNode)
			}
		}
	}

	// Deserialize v1.2 diagnostics
	if diagsRaw, ok := data["diagnostics"].([]any); ok {
		for _, d := range diagsRaw {
			if dMap, ok := d.(map[string]any); ok {
				level, _ := dMap["level"].(string)
				role, _ := dMap["role"].(string)
				message, _ := dMap["message"].(string)
				code, _ := dMap["code"].(string)
				node.Diagnostics = append(node.Diagnostics, NodeDiagnostic{
					Level:   level,
					Role:    role,
					Message: message,
					Code:    code,
				})
			}
		}
	}

	// Deserialize v1.2 annotations
	if anns := deserializeAnnotations(data["annotations"]); anns != nil {
		node.Annotations = anns
	}

	// Deserialize v1.2 catchBranch
	if catchData, ok := data["catchBranch"].([]any); ok {
		for _, c := range catchData {
			if cMap, ok := c.(map[string]any); ok {
				cNode, err := FromJSON(cMap)
				if err != nil {
					return nil, err
				}
				node.CatchBranch = append(node.CatchBranch, *cNode)
			}
		}
	}

	// Deserialize v1.2 finallyBranch
	if finallyData, ok := data["finallyBranch"].([]any); ok {
		for _, f := range finallyData {
			if fMap, ok := f.(map[string]any); ok {
				fNode, err := FromJSON(fMap)
				if err != nil {
					return nil, err
				}
				node.FinallyBranch = append(node.FinallyBranch, *fNode)
			}
		}
	}

	// Deserialize v1.2 asyncVariant
	if av, ok := data["asyncVariant"].(string); ok {
		if av == "all" || av == "race" {
			node.AsyncVariant = av
		}
	}

	// Deserialize v1.2 asyncBody
	if asyncData, ok := data["asyncBody"].([]any); ok {
		for _, a := range asyncData {
			if aMap, ok := a.(map[string]any); ok {
				aNode, err := FromJSON(aMap)
				if err != nil {
					return nil, err
				}
				node.AsyncBody = append(node.AsyncBody, *aNode)
			}
		}
	}

	// Deserialize v1.2 arms
	if armsRaw, ok := data["arms"].([]any); ok {
		for _, ar := range armsRaw {
			if armMap, ok := ar.(map[string]any); ok {
				patternRaw, hasPattern := armMap["pattern"].(map[string]any)
				if !hasPattern {
					continue
				}
				pattern := convertJSONValue(patternRaw)
				var armBody []SemanticNode
				if armBodyData, ok := armMap["body"].([]any); ok {
					for _, b := range armBodyData {
						if bMap, ok := b.(map[string]any); ok {
							bNode, err := FromJSON(bMap)
							if err != nil {
								return nil, err
							}
							armBody = append(armBody, *bNode)
						}
					}
				}
				node.Arms = append(node.Arms, MatchArm{Pattern: pattern, Body: armBody})
			}
		}
	}

	// Deserialize v1.2 defaultArm
	if defaultData, ok := data["defaultArm"].([]any); ok {
		for _, d := range defaultData {
			if dMap, ok := d.(map[string]any); ok {
				dNode, err := FromJSON(dMap)
				if err != nil {
					return nil, err
				}
				node.DefaultArm = append(node.DefaultArm, *dNode)
			}
		}
	}

	return node, nil
}

// deserializeAnnotations converts a raw JSON annotations array to []Annotation.
// Returns nil if the input is nil, not an array, or empty.
func deserializeAnnotations(raw any) []Annotation {
	arr, ok := raw.([]any)
	if !ok || len(arr) == 0 {
		return nil
	}
	var result []Annotation
	for _, item := range arr {
		m, ok := item.(map[string]any)
		if !ok {
			continue
		}
		name, ok := m["name"].(string)
		if !ok || name == "" {
			continue
		}
		ann := Annotation{Name: name}
		if v, ok := m["value"].(string); ok {
			ann.Value = v
		}
		result = append(result, ann)
	}
	if len(result) == 0 {
		return nil
	}
	return result
}

// IsEnvelope returns true if data is an LSEEnvelope (has lseVersion + nodes).
func IsEnvelope(data map[string]any) bool {
	_, hasVersion := data["lseVersion"].(string)
	_, hasNodes := data["nodes"]
	return hasVersion && hasNodes
}

// FromEnvelopeJSON converts a versioned envelope map to an LSEEnvelope.
func FromEnvelopeJSON(data map[string]any) (*LSEEnvelope, error) {
	lseVersion, _ := data["lseVersion"].(string)
	env := &LSEEnvelope{LSEVersion: lseVersion}
	if featuresRaw, ok := data["features"].([]any); ok {
		for _, f := range featuresRaw {
			if s, ok := f.(string); ok {
				env.Features = append(env.Features, s)
			}
		}
	}
	if nodesRaw, ok := data["nodes"].([]any); ok {
		for _, n := range nodesRaw {
			if nMap, ok := n.(map[string]any); ok {
				node, err := FromJSON(nMap)
				if err != nil {
					return nil, err
				}
				env.Nodes = append(env.Nodes, *node)
			}
		}
	}
	return env, nil
}

// ToEnvelopeJSON converts an LSEEnvelope to a JSON-compatible map.
func ToEnvelopeJSON(env *LSEEnvelope) map[string]any {
	result := map[string]any{
		"lseVersion": env.LSEVersion,
		"nodes":      []any{},
	}
	if len(env.Features) > 0 {
		result["features"] = env.Features
	}
	nodeMaps := make([]any, len(env.Nodes))
	for i := range env.Nodes {
		nodeMaps[i] = ToJSON(&env.Nodes[i])
	}
	result["nodes"] = nodeMaps
	return result
}

// ToJSON converts a SemanticNode to a full-fidelity JSON-friendly map.
func ToJSON(node *SemanticNode) map[string]any {
	m := map[string]any{
		"kind":   string(node.Kind),
		"action": node.Action,
		"roles":  marshalRoles(node.Roles),
	}
	if node.Kind == KindEventHandler && len(node.Body) > 0 {
		bodyMaps := make([]any, len(node.Body))
		for i := range node.Body {
			bodyMaps[i] = ToJSON(&node.Body[i])
		}
		m["body"] = bodyMaps
	}
	if node.Kind == KindCommand && len(node.Body) > 0 {
		bodyMaps := make([]any, len(node.Body))
		for i := range node.Body {
			bodyMaps[i] = ToJSON(&node.Body[i])
		}
		m["body"] = bodyMaps
	}
	if node.Kind == KindCompound {
		stmtMaps := make([]any, len(node.Statements))
		for i := range node.Statements {
			stmtMaps[i] = ToJSON(&node.Statements[i])
		}
		m["statements"] = stmtMaps
		if node.ChainType != "" {
			m["chainType"] = node.ChainType
		}
	}
	// v1.1 conditional fields
	if len(node.ThenBranch) > 0 {
		maps := make([]any, len(node.ThenBranch))
		for i := range node.ThenBranch {
			maps[i] = ToJSON(&node.ThenBranch[i])
		}
		m["thenBranch"] = maps
	}
	if len(node.ElseBranch) > 0 {
		maps := make([]any, len(node.ElseBranch))
		for i := range node.ElseBranch {
			maps[i] = ToJSON(&node.ElseBranch[i])
		}
		m["elseBranch"] = maps
	}
	// v1.1 loop fields
	if node.LoopVariant != "" {
		m["loopVariant"] = node.LoopVariant
	}
	if len(node.LoopBody) > 0 {
		maps := make([]any, len(node.LoopBody))
		for i := range node.LoopBody {
			maps[i] = ToJSON(&node.LoopBody[i])
		}
		m["loopBody"] = maps
	}
	if node.LoopVariable != "" {
		m["loopVariable"] = node.LoopVariable
	}
	if node.IndexVariable != "" {
		m["indexVariable"] = node.IndexVariable
	}
	// v1.2 fields
	if len(node.Diagnostics) > 0 {
		diags := make([]any, len(node.Diagnostics))
		for i, d := range node.Diagnostics {
			diags[i] = map[string]any{
				"level":   d.Level,
				"role":    d.Role,
				"message": d.Message,
				"code":    d.Code,
			}
		}
		m["diagnostics"] = diags
	}
	if len(node.Annotations) > 0 {
		anns := make([]any, len(node.Annotations))
		for i, a := range node.Annotations {
			if a.Value != "" {
				anns[i] = map[string]any{"name": a.Name, "value": a.Value}
			} else {
				anns[i] = map[string]any{"name": a.Name}
			}
		}
		m["annotations"] = anns
	}
	if len(node.CatchBranch) > 0 {
		maps := make([]any, len(node.CatchBranch))
		for i := range node.CatchBranch {
			maps[i] = ToJSON(&node.CatchBranch[i])
		}
		m["catchBranch"] = maps
	}
	if len(node.FinallyBranch) > 0 {
		maps := make([]any, len(node.FinallyBranch))
		for i := range node.FinallyBranch {
			maps[i] = ToJSON(&node.FinallyBranch[i])
		}
		m["finallyBranch"] = maps
	}
	if node.AsyncVariant != "" {
		m["asyncVariant"] = node.AsyncVariant
	}
	if len(node.AsyncBody) > 0 {
		maps := make([]any, len(node.AsyncBody))
		for i := range node.AsyncBody {
			maps[i] = ToJSON(&node.AsyncBody[i])
		}
		m["asyncBody"] = maps
	}
	if len(node.Arms) > 0 {
		arms := make([]any, len(node.Arms))
		for i, arm := range node.Arms {
			bodyMaps := make([]any, len(arm.Body))
			for j := range arm.Body {
				bodyMaps[j] = ToJSON(&arm.Body[j])
			}
			arms[i] = map[string]any{
				"pattern": marshalValue(arm.Pattern),
				"body":    bodyMaps,
			}
		}
		m["arms"] = arms
	}
	if len(node.DefaultArm) > 0 {
		maps := make([]any, len(node.DefaultArm))
		for i := range node.DefaultArm {
			maps[i] = ToJSON(&node.DefaultArm[i])
		}
		m["defaultArm"] = maps
	}
	return m
}

// convertJSONValue converts a JSON value object to a SemanticValue.
func convertJSONValue(data map[string]any) SemanticValue {
	vtype, _ := data["type"].(string)

	switch vtype {
	case "selector":
		val, _ := data["value"].(string)
		sv := SelectorValue(val)
		if sk, ok := data["selectorKind"].(string); ok {
			sv.SelectorKind = sk
		}
		return sv

	case "literal":
		value := data["value"]
		switch v := value.(type) {
		case bool:
			return LiteralValue(v, "boolean")
		case float64:
			// JSON numbers are always float64
			if v == float64(int64(v)) {
				return LiteralValue(int(v), "number")
			}
			return LiteralValue(v, "number")
		default:
			dataType, _ := data["dataType"].(string)
			if dataType == "" {
				dataType = "string"
			}
			s := fmt.Sprintf("%v", value)
			return LiteralValue(s, dataType)
		}

	case "reference":
		val := fmt.Sprintf("%v", data["value"])
		return ReferenceValue(val)

	case "expression":
		raw, _ := data["raw"].(string)
		if raw == "" {
			raw = fmt.Sprintf("%v", data["value"])
		}
		return ExpressionValue(raw)

	case "flag":
		name, _ := data["name"].(string)
		if name == "" {
			name = fmt.Sprintf("%v", data["value"])
		}
		enabled := true
		if e, ok := data["enabled"].(bool); ok {
			enabled = e
		}
		return FlagValue(name, enabled)

	case "property-path":
		val := fmt.Sprintf("%v", data["value"])
		return ExpressionValue(val)

	default:
		val := fmt.Sprintf("%v", data["value"])
		return LiteralValue(val, "string")
	}
}
